src/storage/src/healthcheck/ssh.rs

```diff
@@ -0,0 +1,124 @@
+
+    let (input, mut channels) = unbounded_channel();
+
+    let button = op.build(|mut caps| async move {
+        let mut capability = Some(caps.pop().unwrap());
+
+        let mut tunnel_statuses = BTreeMap::new();
+        let mut subscriptions = tokio_stream::StreamMap::new();
+
+        loop {
+            let closed = matches!(
+                channels.try_recv(),
+                Err(tokio::sync::mpsc::error::TryRecvError::Disconnected)
+            );
+            let no_subscriptions = subscriptions.is_empty();
+
+            if closed && no_subscriptions {
+                break;
+            }
+
+            tokio::select! {
+                Some((key, channel)) = channels.recv(), if !closed => {
+                    // Add a new subscription to the `StreamMap`. We prioritize newer
+                    // subscriptions for the same key, but in practice the channels are
+                    // the same.
+                    subscriptions.insert(key, channel);
+                },
+                Some((key, status)) = subscriptions.next(), if !no_subscriptions => {
+                    tunnel_statuses.insert(key, status);
+
+                    let mut status_to_emit = HealthStatusMessage {
+                        index: 0,
+                        namespace: StatusNamespace::Ssh,
+                        update: HealthStatusUpdate::Running
+                    };
+
+                    // Create a single `Ssh`-namespaced errors out of all the different ssh tunnels
+                    // involved in this source.
+                    //
+                    // In practice, this is exactly one, but until we deprecate per-broker ssh
+                    // tunnels, we have to support multiple.
+                    for (_key, status) in tunnel_statuses.iter() {
+                        match (&mut status_to_emit.update, status) {
+                            (HealthStatusUpdate::Running, SshTunnelStatus::Errored(e)) => {
+                                status_to_emit.update = HealthStatusUpdate::Stalled {
+                                    error: e.to_string(),
+                                    hint: None,
+                                    should_halt: false
+                                };
+                            }
+                            (
+                                HealthStatusUpdate::Stalled {
+                                    ref mut error,
+                                    ..
+                                },
+                                SshTunnelStatus::Errored(e)
+                            ) => {
+                                // The inference engine really doesn't like this method called any
+                                // other way!
+                                error.push_str(&format!(", {}", ErrorExt::display_with_causes(&**e)));
+                            }
+                            _ => {}
+                        }
+                    }
```

Comment:
From a maintenance perspective, this looks like rather deeply nested code. Any chances of extracting a method here? (optional)

Reply:
I actually think cause its a timely operator it would make it harder to read to extract it!